src/main/java/org/jenkinsci/plugins/ssegateway/EventHistoryStore.java

```diff
@@ -0,0 +1,267 @@
+        if (channelDir == null) {
+            channelDir = new File(historyRoot, channelName);
+            channelDirs.put(channelName, channelDir);
+        }
+        if (!channelDir.exists()) {
+            if (!channelDir.mkdirs()) {
+                throw new IOException(String.format("Unexpected error creating channel event log dir %s.", channelDir.getAbsolutePath()));
+            }
+        }
+        
+        return channelDir;
+    }
+    
+    private synchronized static void deleteAllFilesInDir(File dir, Long olderThan) throws IOException {
+        Path dirPath = Paths.get(dir.toURI());
+        
+        try (final DirectoryStream<Path> dirStream = Files.newDirectoryStream(dirPath)) {
+            for (final Path entry : dirStream) {
+                File file = entry.toFile();
+                if (file.isDirectory()) {
+                    deleteAllFilesInDir(file, olderThan);
+                }
+                if (olderThan == null || file.lastModified() < olderThan) {
+                    if (!file.delete()) {
+                        LOGGER.log(Level.SEVERE, "Error deleting file " + file.getAbsolutePath());
+                    }
+                }
+            }
+        }
+    }
+
+    private static void assertHistoryRootSet() {
+        if (historyRoot == null) {
+            throw new IllegalStateException("'historyRoot' not set. Check for earlier initialization errors.");
+        }
+    }
+    
+    public synchronized static void enableAutoDeleteOnExpire() {
+        if (autoExpireTimer != null) {
+            return;
+        }
+        
+        // Set up a timer that runs the DeleteStaleHistoryTask 3 times over
+        // duration of the event expiration timeout. By default this will be
+        // every 20 seconds i.e. in that case, events are never left lying around
+        // for more than 20 seconds minute past their expiration.
+        long taskSchedule = expiresAfter / 3;
+        autoExpireTimer = new Timer();
+        autoExpireTimer.schedule(new DeleteStaleHistoryTask(), taskSchedule, taskSchedule);
+        Runtime.getRuntime().addShutdownHook(new Thread() {
+            @Override
+            public void run() {
+                disableAutoDeleteOnExpire();
+            }
+        });
+    }
+    
+    public synchronized static void disableAutoDeleteOnExpire() {
+        if (autoExpireTimer == null) {
+            return;
+        }
+        autoExpireTimer.cancel();
+        autoExpireTimer = null;
+    }
+    
+    private static class DeleteStaleHistoryTask extends TimerTask {
+        @Override
+        public void run() {
+            try {
+                deleteStaleHistory();
+            } catch (Exception e) {
+                LOGGER.log(Level.SEVERE, "Error deleting stale/expired events from EventHistoryStore.", e);
+            }
+        }
+    }
+    
+    private static synchronized AtomicInteger getChannelSubsCounter(@Nonnull String channelName) {
```

Comment:
Not sure if it needs synchronization, calling method onChannelSubscribe() is synchronized anyways?